lib/generateNormals.js

```diff
@@ -20,24 +40,173 @@ module.exports = generateNormals;
  * @see addPipelineExtras
  * @see loadGltfUris
  */
-function generateNormals(gltf) {
+function generateNormals(gltf, options) {
+    options = defaultValue(options, {});
+    var faceNormals = defaultValue(options.faceNormals, false);
     var primitives = getPrimitives(gltf);
     var primitivesLength = primitives.length;
     for (var i = 0; i < primitivesLength; i++) {
         var primitive = primitives[i];
-        var positionAccessorId = primitive.attributes.POSITION;
-        var normalAccessorId = primitive.attributes.NORMAL;
-        var indicesAccessorId = primitive.indices;
-
-        if (primitive.mode === WebGLConstants.TRIANGLES && defined(indicesAccessorId) && defined(positionAccessorId) && !defined(normalAccessorId)) {
-            var geometry = gltfPrimitiveToCesiumGeometry(gltf, primitive);
-            GeometryPipeline.computeNormal(geometry);
-            cesiumGeometryToGltfPrimitive(gltf, primitive, geometry);
+        var positionSemantics = getPrimitiveAttributeSemantics(primitive, 'POSITION');
+        var normalSemantics = getPrimitiveAttributeSemantics(primitive, 'NORMAL');
+        var generatedMaterials = {};
+        if (primitive.mode === WebGLConstants.TRIANGLES && positionSemantics.length > 0 && normalSemantics.length === 0) {
+            if (faceNormals) {
+                generateFaceNormals(gltf, primitive, positionSemantics[0]);
+            } else {
+                var geometry = gltfPrimitiveToCesiumGeometry(gltf, primitive);
+                GeometryPipeline.computeNormal(geometry);
+                cesiumGeometryToGltfPrimitive(gltf, primitive, geometry);
+            }
+            generateMaterial(gltf, primitive, generatedMaterials);
         }
     }
+
+    removeAccessors(gltf);
+    removeBufferViews(gltf);
+    removeBuffers(gltf);
+    mergeBuffers(gltf);
+    mergeDuplicateVertices(gltf);
```

Comment:
Since `mergeDuplicateVertices` is the next step in `Pipeline` we can remove it from here.